Src/AutoFixture/ElementsBuilder.cs

```diff
@@ -0,0 +1,57 @@
+﻿using Ploeh.AutoFixture.Kernel;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Ploeh.AutoFixture
+{
+    /// <summary>
+    /// Draws a random element from the given collection
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public sealed class ElementsBuilder<T> : ISpecimenBuilder
+    {
+        private readonly ICollection<T> elements;
+        private readonly RandomNumericSequenceGenerator sequence;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ElementsBuilder{T}"/> class.
+        /// </summary>
+        /// <param name="elements">The collection from which elements should be drawn from.
+        /// It must contain at least two element.</param>
+        public ElementsBuilder(ICollection<T> elements)
```

Comment:
I grant that calling `Count()` could have side-effects, if traversing the sequence produces side-effects. It shouldn't, but who knows what a user might do...

I'm happy taking on that risk, and essentially declaring that side-effecty sequences constitute a user error.

Using any sort of mutable collection comes with the risk of side-effects as well, in the sense that a user could pass in a `List<T>` and then subsequently modify the collection. Essentially, the 'not empty' constructor Guard Clause is no guarantee, because the user could empty the collection after the constructor runs:

```C#
var l = new List<int> { 1, 2, 3 };
var b = new ElementsBuilder<int>(l);
l.Clear();
``` 

FWIW, the [most appropriate type to use is `IReadOnlyCollection<T>`](http://blog.ploeh.dk/2013/07/20/linq-versus-the-lsp), but unfortunately, that's only available on .NET 4.5+.

Another option is to use an array, but arrays are also 'slightly mutable'.

In AutoFixture, the most idiomatic type to use would be `IEnumerable<T>`, despite the flawed guarantees it gives.